src/ucp/rndv/proto_rndv.c

```diff
@@ -1081,3 +1194,74 @@ void ucp_proto_rndv_bulk_request_init_lane_idx(
 
     req->send.multi_lane_idx = lane_idx - 1;
 }
+
+static int
+ucp_proto_rndv_predict_rtr_mtype(const ucp_proto_init_params_t *init_params)
+{
+    ucp_context_t *context          = init_params->worker->context;
+    ucs_memory_type_t frag_mem_type = context->config.ext.rndv_frag_mem_type;
+
+    if (init_params->rkey_config_key == NULL) {
+        return 0;
+    }
+
+    return (init_params->rkey_config_key->mem_type == frag_mem_type) ? 1 : 0;
+}
+
+ucs_status_t
+ucp_proto_rndv_predict_prev_stages(const ucp_proto_init_params_t *init_params,
+                                   ucp_proto_caps_t *caps)
+{
+    int is_rtr_mtype  = ucp_proto_rndv_predict_rtr_mtype(init_params);
+    size_t num_stages = 0;
+    ucp_proto_perf_range_t ctrl_stages[2];
+    ucp_proto_rndv_ctrl_init_params_t rts_params, rtr_params;
+    ucp_md_map_t remote_md_map, local_md_map;
+    ucp_proto_rndv_ctrl_priv_t rpriv;
+    ucp_proto_caps_t tmp_caps;
+    ucs_status_t status;
+
+    ucp_proto_rndv_rts_params_init(init_params, &rts_params);
+    if (rts_params.lane == UCP_NULL_LANE) {
+        return UCS_ERR_NO_ELEM;
+    }
+
+    /* Init RTS priv */
+    ucp_proto_rndv_ctrl_init_priv(&rts_params, &rpriv, rts_params.lane);
+    local_md_map  = rpriv.md_map;
+    remote_md_map = ucp_proto_rndv_md_map_to_remote(&rts_params, rpriv.md_map);
+
+    /* Turn RTS md_map to remote in case if RTS was sent by peer */
+    if (ucp_proto_init_check_op(init_params, UCS_BIT(UCP_OP_ID_RNDV_SEND))) {
+        rpriv.md_map = remote_md_map;
+    }
+    status = ucp_proto_rndv_ctrl_range_init(&rts_params, &rpriv,
+                                            &ctrl_stages[num_stages++]);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    /* Predict RTR */
+    if (ucp_proto_init_check_op(init_params, UCS_BIT(UCP_OP_ID_RNDV_SEND))) {
+        if (is_rtr_mtype) {
+            status = ucp_proto_rndv_rtr_mtype_params_init(init_params,
+                                                          &rtr_params);
+        } 
+        if ((!is_rtr_mtype) || (status != UCS_OK)) {
+            ucp_proto_rndv_rtr_params_init(init_params, &rtr_params);
+        }
+
```

Comment:
maybe
```suggestion
        if (ucp_proto_rndv_predict_rtr_mtype(init_params)) {
            status = ucp_proto_rndv_rtr_mtype_params_init(init_params,
                                                          &rtr_params);
            ucs_assertv_always(status == UCS_OK, "status %s", ucs_status_string(status));
        } else {
            ucp_proto_rndv_rtr_params_init(init_params, &rtr_params);
        }

```

and remove `is_rtr_mtype`

Reply:
I don't think that this is good idea since in that patch there is no way to certainly predict when does `rtr/mtype` happens. 

So it seems like `rtr_mtype_params_init` can fail (e.g. when `worker->mem_type_ep[mem_type] == NULL` and I think that the best option in that case is to fall back to usual RTR message prediction.